packages/pricing/src/services/pricing-module.ts

```diff
@@ -79,46 +97,90 @@ export default class PricingModuleService<
+    // that the price set is configured for
+    const psmaSubQueryKnex = knex({
+      psma: "price_set_money_amount",
+    })
+      .select({
+        id: "psma.id",
+        price_set_id: "psma.price_set_id",
+        money_amount_id: "psma.money_amount_id",
+        number_rules: "psma.number_rules",
+      })
+      .join("price_rule as pr", "pr.price_set_money_amount_id", "psma.id")
+      .join("rule_type as rt", "rt.id", "pr.rule_type_id")
+      .orderBy("number_rules", "desc")
+
+    for (const [key, value] of Object.entries(context)) {
+      psmaSubQueryKnex.orWhere({
+        "rt.rule_attribute": key,
+        "pr.value": value,
+      })
     }
 
-    const priceSets = await this.list(
-      priceSetFilters,
-      {
-        select: [
-          "id",
-          "money_amounts.id",
-          "money_amounts.currency_code",
-          "money_amounts.amount",
-          "money_amounts.min_quantity",
-          "money_amounts.max_quantity",
-        ],
-        relations: ["money_amounts"],
-      },
-      sharedContext
-    )
+    psmaSubQueryKnex
+      .groupBy("psma.id")
+      .having(knex.raw("count(DISTINCT rt.rule_attribute) = psma.number_rules"))
+      .andHaving(knex.raw(`${Object.entries(context).length} > 0`))
 
-    const calculatedPrices = priceSets.map(
-      (priceSet): PricingTypes.CalculatedPriceSetDTO => {
-        // TODO: This will change with the rules engine selection,
-        // making a DB query directly instead
-        // This should look for a default price when no rules apply
-        // When no price is set, return null values for all cases
-        const selectedMoneyAmount = priceSet.money_amounts?.find(
-          (ma) =>
-            context.currency_code && ma.currency_code === context.currency_code
-        )
+    const priceSetQueryKnex = knex({
+      ps: "price_set",
+    })
+      .select({
+        id: "ps.id",
+        amount: "ma.amount",
+        min_quantity: "ma.min_quantity",
+        max_quantity: "ma.max_quantity",
+        currency_code: "ma.currency_code",
+        default_priority: "rt.default_priority",
+        number_rules: "psma.number_rules",
+      })
+      .join(psmaSubQueryKnex.as("psma"), "psma.price_set_id", "ps.id")
+      .join("money_amount as ma", "ma.id", "psma.money_amount_id")
+      .join("price_rule as pr", "pr.price_set_money_amount_id", "psma.id")
+      .join("rule_type as rt", "rt.id", "pr.rule_type_id")
+      .whereIn("ps.id", pricingFilters.id)
+      .orderBy([
+        { column: "number_rules", order: "desc" },
+        { column: "default_priority", order: "desc" },
+      ])
+
+    if (quantity) {
+      priceSetQueryKnex.where("ma.min_quantity", "<=", quantity)
+      priceSetQueryKnex.andWhere("ma.max_quantity", ">=", quantity)
+    }
+
+    const queryBuilderResults = await priceSetQueryKnex
+    const groupedPricesByPriceSetId = groupBy(queryBuilderResults, "id")
```

Comment:
I believe we should move only the functions we use from `lodash` to the repo. The package is quite big to have it the module.

for this groupBy we could just wrap the query like that:
```sql
SELECT 
  price_set_id, 
  json_agg(row_to_json(X)) AS prices
FROM (
  --query
) X
GROUP BY price_set_id;
```
This is just a suggestion though. The number of rows is small and won't impact the performance

Reply:
I still want to keep this js based for now, I've implemented the groupBy in the utils.